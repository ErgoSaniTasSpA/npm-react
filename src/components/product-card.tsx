import React,{ createContext, CSSProperties  } from 'react';

import styles from '../styles/styles.module.css';

import { UseProduct } from '../hooks/use-product';
import { OnChangeArgs, Product, ProductContextProps,InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext( {} as ProductContextProps);
const { Provider } = ProductContext;


export interface ProductProps {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: ( args: ProductCardHandlers ) => JSX.Element
  className?: string;
  style?: CSSProperties;
  onChange?: ( args : OnChangeArgs ) => void;
  value?: number;
  initialValues?: InitialValues
}

/**************************** NOTA IMPORTANTE ************************************************
* * Provider : Envuelve la aplicación React y se encarga de suministrar el STORE de REDUX a todos 
* * los componentes que se encuentren dentro de la aplicación. De esta manera, cualquier componente 
* * que necesite acceder al STORE (ProductContextProps) puede hacerlo utilizando los hooks 
* * o funciones específicas proporcionadas por React-Redux.
* *********************************************************************************************/

export const ProductCard = ( {children,product,className,style,onChange,value,initialValues }:ProductProps) => {

  const { counter , increaseBy,isMaxCountReached,maxCount,reset } = 
        UseProduct( { onChange , product, value ,initialValues } );

  return (

    <Provider value = {{ counter,increaseBy,product,maxCount }} >
      {/* Aqui va Body del page */}
      <div className={`${ styles.productCard } ${ className }`} style= { style } >
        {/*  Ejecuta una funcion JSX Element (jsx) renderizable, SI retorna parametros */}
        { 
          children({
            counter,
            isMaxCountReached,
            maxCount,
            product,
            increaseBy,
            reset
          })
        } 
      </div>
    
    </Provider>
  )
}
