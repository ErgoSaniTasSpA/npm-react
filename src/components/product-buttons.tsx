import React,{ CSSProperties, useCallback, useContext } from "react";
import { ProductContext } from "./product-card";
import styles from '../styles/styles.module.css';

export interface Props {
    className? : string;
    style?: CSSProperties;
}

export const ProductButtons = ( { className,style } : Props ) => {

    const { counter,increaseBy,maxCount  } = useContext(ProductContext);


    //Retorna una funcion
    const isMaxReached = useCallback( 
      // '!!maxCount' => Si existe (true) lo transforma a boolea true/false
      () => !!maxCount && counter === maxCount,
    [counter,maxCount],
  );
  
    return (
      <div 
        style = { style }
        className={` ${styles.buttonsContainer } ${className} `} >
          <button
              className={ styles.buttonMinus } 
              onClick={ () => increaseBy ( -1 ) } > - </button>
          <button
              className={ styles.countLabel } 
              onClick={ () => increaseBy ( counter ) } > { counter } </button>
          <button
              className={ `${styles.buttonAdd}  ${ isMaxReached() && styles.disabled }` } 
              onClick={ () => increaseBy ( +1 ) } > + </button>        
      </div>
    )
  }