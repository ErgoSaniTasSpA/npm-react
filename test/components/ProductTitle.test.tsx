

import * as React from 'react';
import  renderer  from 'react-test-renderer';
import { ProductCard, ProductTitle } from '../../src/components';
import { products } from './data/products';
/*
* Instalar
* - npm i react-test-renderer
* - npm add -D identity-obj-proxy
*/

describe('TEST Component ProductTitle', () => {

    test('Visualizar Component Product Titulo', () => {

    
        const wrapper = renderer.create(
            <ProductTitle title='Test Custom Product' />
        )
        expect (wrapper.toJSON()).toMatchSnapshot();
    });

    test('Visualizar Product Card con Product Title', () => {

    
        const wrapper = renderer.create(
            <ProductCard product= { products[0] } >
            {
                () => ( <ProductTitle /> )
            }    
            </ProductCard>     
        )
        expect (wrapper.toJSON()).toMatchSnapshot();
    });
    


});