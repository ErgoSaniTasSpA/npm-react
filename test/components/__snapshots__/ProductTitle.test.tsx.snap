// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST Component ProductTitle Visualizar Component Product Titulo 1`] = `
<span
  className=" productDescription undefined "
>
   
  Test Custom Product
</span>
`;

exports[`TEST Component ProductTitle Visualizar Product Card con Product Title 1`] = `
<div
  className="productCard undefined"
>
  <span
    className=" productDescription undefined "
  >
     
    Coffe Mug
  </span>
</div>
`;
